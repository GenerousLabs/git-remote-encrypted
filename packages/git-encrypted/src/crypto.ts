import { arrayToHex, hexToArray } from 'enc-utils';
import { hash, randomBytes, secretbox } from 'tweetnacl';
import { decodeUTF8, encodeBase64, encodeUTF8 } from 'tweetnacl-util';
import { NONCE_LENGTH } from './constants';
import { Keys, GitBaseParamsEncrypted } from './types';
import { doesDirectoryExist } from './utils';
import { deriveKeys } from './keyDerivation';
import { saveKeysToDisk } from './api/saveKeysToDisk';
import { ensureMetaExists } from './encryptedMeta';

export const concat = (a: Uint8Array, b: Uint8Array) => {
  const length = a.length + b.length;
  const result = new Uint8Array(length);
  result.set(a, 0);
  result.set(b, a.length);
  return result;
};

export const split = (combined: Uint8Array, length: number) => {
  const a = combined.slice(0, length);
  const b = combined.slice(length);
  return [a, b];
};

export const generateKey = (length: number) =>
  encodeBase64(randomBytes(length));

/**
 * Given an input and a salt, combine them into an nonce of the correct length.
 * This is a deterministic operation.
 *
 * NOTE: Any feedback from cryptographers welcome here. This might be a
 * terrible idea.
 */
export const createNonce = async ({
  salt,
  input,
}: {
  salt: Uint8Array;
  input: Uint8Array;
}) => {
  const combined = concat(input, salt);
  const hashed = hash(combined);
  const [nonce] = split(hashed, NONCE_LENGTH);
  return nonce;
};

// Copy the same API as secretbox
export const encrypt = async (
  msg: Uint8Array,
  nonce: Uint8Array,
  key: Uint8Array
) => {
  const result = secretbox(msg, nonce, key);
  return result;
};

// Copy the same API as secretbox.open
export const decrypt = async (
  box: Uint8Array,
  nonce: Uint8Array,
  key: Uint8Array
) => {
  const decrypted = secretbox.open(box, nonce, key);
  if (decrypted === null) {
    throw new Error('Decryption failed. #FxljFi');
  }
  return decrypted;
};

/**
 * Given a cleartext filename, calculate the encrypted filename
 *
 * NOTE: This cannot be decrypted. The nonce will be lost. `encryptFile()`
 * saves the nonce into the file body.
 */
export const encryptFilename = async ({
  filenameArray,
  keys,
  nonce,
}: {
  filenameArray: Uint8Array;
  nonce: Uint8Array;
  keys: Keys;
}) => {
  const encryptedFilenameArray = await encrypt(
    filenameArray,
    nonce,
    keys.filename
  );
  const encryptedFilenameHex = arrayToHex(encryptedFilenameArray);
  return encryptedFilenameHex;
};

/**
 * Given a cleartext filename and a contents, create an nonce then use that to
 * encrypt the filename, encrypt the file contents, and combine the nonce and
 * encrypted file contents, returning both the encrypted filename (hex encoded)
 * and the encrypted body combined with the nonce.
 */
export const encryptFile = async ({
  filename,
  contents,
  keys,
}: {
  filename: string;
  contents: Uint8Array;
  keys: Keys;
}): Promise<[string, Uint8Array]> => {
  const filenameArray = decodeUTF8(filename);
  const nonce = await createNonce({ salt: keys.salt, input: filenameArray });

  const encryptedFilenameHex = await encryptFilename({
    filenameArray,
    keys,
    nonce,
  });

  const encryptedContents = await encrypt(contents, nonce, keys.content);
  const noncePlusEncryptedContent = concat(nonce, encryptedContents);

  return [encryptedFilenameHex, noncePlusEncryptedContent];
};

export const decryptFile = async ({
  fileContents,
  encryptedFilenameHex,
  keys,
}: {
  encryptedFilenameHex: string;
  fileContents: Uint8Array;
  keys: Keys;
}): Promise<[string, Uint8Array]> => {
  const [nonce, encryptedContents] = split(fileContents, NONCE_LENGTH);

  const encryptedFilenameArray = hexToArray(encryptedFilenameHex);
  const decryptedFilenameArray = await decrypt(
    encryptedFilenameArray,
    nonce,
    keys.filename
  );
  const decryptedFilename = encodeUTF8(decryptedFilenameArray);

  const decryptedContents = await decrypt(
    encryptedContents,
    nonce,
    keys.content
  );

  return [decryptedFilename, decryptedContents];
};

export const decryptFileContentsOnly = async ({
  fileContents,
  keys,
}: {
  fileContents: Uint8Array;
  keys: Keys;
}) => {
  const [nonce, encryptedContents] = split(fileContents, NONCE_LENGTH);

  const decryptedContents = await decrypt(
    encryptedContents,
    nonce,
    keys.content
  );

  return decryptedContents;
};

export const ensureKeysExist = async ({
  fs,
  gitdir,
  encryptedDir,
  encryptedKeysDir,
  keyDerivationPassword,
}: Pick<GitBaseParamsEncrypted, 'fs' | 'gitdir'> & {
  encryptedDir: string;
  encryptedKeysDir: string;
  keyDerivationPassword?: string;
}) => {
  const encryptedKeysDirectoryExists = await doesDirectoryExist({
    fs,
    path: encryptedKeysDir,
  });

  if (encryptedKeysDirectoryExists) {
    return;
  }

  if (typeof keyDerivationPassword === 'undefined') {
    throw new Error(
      'Cannot initialise without key derivation password #xvoEqa'
    );
  }

  const meta = await ensureMetaExists({ fs, encryptedDir });

  const keys = await deriveKeys({ ...meta.derivationParams, password: keyDerivationPassword });
  await saveKeysToDisk({ fs, gitdir, keys });
};