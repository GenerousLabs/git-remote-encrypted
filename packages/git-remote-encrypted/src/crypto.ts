import { arrayToHex } from 'enc-utils';
import { hash, randomBytes, secretbox } from 'tweetnacl';
import { decodeBase64, decodeUTF8, encodeBase64 } from 'tweetnacl-util';

export type KEYS = {
  secret: Uint8Array;
  secretNonce: Uint8Array;
  filenames: Uint8Array;
  filenamesNonce: Uint8Array;
};
const DEFAULT_KEYS: KEYS = {
  secret: decodeBase64('OTdY2G5jOtUb4NkTIrcMic5Om2FSGVNr+mOV21bMfkY='),
  secretNonce: decodeBase64('wIfKspQFPMhcpxWSNO/d/aA50ErheC6t'),
  filenames: decodeBase64('aeFYzTwOrPbAPu7Lyw1QZ34JglphbLTgAAHtjr2Zcps='),
  filenamesNonce: decodeBase64('JbTmJEJIT3fx2agUFmLFkb0Zk60/Eeoa'),
};

const { nonceLength: NONCE_LENGTH } = secretbox;

export const concat = (a: Uint8Array, b: Uint8Array) => {
  const length = a.length + b.length;
  const result = new Uint8Array(length);
  result.set(a, 0);
  result.set(b, a.length);
  return result;
};

export const split = (combined: Uint8Array, length: number) => {
  const a = combined.slice(0, length);
  const b = combined.slice(length);
  return [a, b];
};

export const generateKey = (length: number) =>
  encodeBase64(randomBytes(length));

export const objectIdToNonce = (objectId: string, nonceKey: Uint8Array) => {
  const combined = concat(decodeUTF8(objectId), nonceKey);
  const hashed = hash(combined);
  const [nonce] = split(hashed, NONCE_LENGTH);
  return nonce;
};

export const encryptFilename = (objectId: string, keys: KEYS) => {
  const nonce = objectIdToNonce(objectId, keys.filenamesNonce);
  const objectIdUint8Array = decodeUTF8(objectId);
  const box = secretbox(objectIdUint8Array, nonce, keys.filenames);
  const output = arrayToHex(box);
  return output;
};

export const encryptContents = (
  objectId: string,
  wrappedContent: Uint8Array,
  keys: KEYS
) => {
  const nonce = objectIdToNonce(objectId, keys.secretNonce);

  const box = secretbox(wrappedContent, nonce, keys.secret);

  return concat(nonce, box);
};

export const encrypt = (
  objectId: string,
  wrappedContent: Uint8Array,
  keys = DEFAULT_KEYS
): [string, Uint8Array] => {
  const filename = encryptFilename(objectId, keys);
  const content = encryptContents(objectId, wrappedContent, keys);
  return [filename, content];
};

// export const decryptFilename = (encryptedFilename: string, keys: KEYS) => {
//   const encryptedFilenameArray = decodeUTF8(encryptFilename);
//   const [nonce, box] = split(encryptedFilenameArray, NONCE_LENGTH);
// };

export const decrypt = (encryptedContent: Uint8Array, keys = DEFAULT_KEYS) => {
  const [nonce, box] = split(encryptedContent, NONCE_LENGTH);

  const wrappedContent = secretbox.open(box, nonce, keys.secret);

  if (wrappedContent === null) {
    throw new Error('Failed to decrypt. #hzpdW0');
  }

  return wrappedContent;
};
