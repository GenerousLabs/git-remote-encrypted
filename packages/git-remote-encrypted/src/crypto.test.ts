import { decodeBase64, decodeUTF8 } from 'tweetnacl-util';
import { encryptFilename, concat, split, decrypt, encrypt } from './crypto';

const keys = {
  secret: decodeBase64('OTdY2G5jOtUb4NkTIrcMic5Om2FSGVNr+mOV21bMfkY='),
  secretNonce: decodeBase64('wIfKspQFPMhcpxWSNO/d/aA50ErheC6t'),
  filenames: decodeBase64('aeFYzTwOrPbAPu7Lyw1QZ34JglphbLTgAAHtjr2Zcps='),
  filenamesNonce: decodeBase64('JbTmJEJIT3fx2agUFmLFkb0Zk60/Eeoa'),
};

const exampleObjectId = '6fb50f57ebfe2687563a016a1cdce2f6ee445312';
const exampleEncryptedFilename =
  '4c0e2aafaf641002e186c1ca75d9d7610dae59cf4224aa89d0a375863af4b3fbeb486ad88d76edda3b7c797d425cb27295bc7a561ab0c0d2';
const exampleString = 'This is a simple text string';
const exampleArray = decodeUTF8(exampleString);

describe('crypto', () => {
  describe('cocnat()', () => {
    it('Successfully combines arrays #UFUHj5', () => {
      expect(
        concat(Uint8Array.from([255, 1, 0, 255]), Uint8Array.from([13]))
      ).toEqual(Uint8Array.from([255, 1, 0, 255, 13]));
    });
  });

  describe('split()', () => {
    it('Successfully splits an array #i2yH7q', () => {
      expect(
        split(
          Uint8Array.from([255, 1, 0, 255, 13]),

          4
        )
      ).toEqual([Uint8Array.from([255, 1, 0, 255]), Uint8Array.from([13])]);
    });
  });

  describe('encryptFilename()', () => {
    it('Encrypts to a known value #0IJkyE', () => {
      expect(encryptFilename(exampleObjectId, keys)).toEqual(
        exampleEncryptedFilename
      );
    });
  });

  describe('encrypt()', () => {
    it('Matches the snapshot #fbz3Op', () => {
      expect(encrypt(exampleObjectId, exampleArray, keys)).toMatchSnapshot();
    });
  });

  describe('encrypt() and decrypt()', () => {
    it('Correctly encrypts and decrypts back to the same #v3VSuD', () => {
      const [, encrypted] = encrypt(exampleObjectId, exampleArray, keys);
      const decrypted = decrypt(encrypted, keys);
      expect(decrypted).toEqual(exampleArray);
    });
  });
});
