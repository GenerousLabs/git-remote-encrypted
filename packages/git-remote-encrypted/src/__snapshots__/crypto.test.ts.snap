// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`crypto encrypt() Matches the snapshot #fbz3Op 1`] = `
Array [
  "4c0e2aafaf641002e186c1ca75d9d7610dae59cf4224aa89d0a375863af4b3fbeb486ad88d76edda3b7c797d425cb27295bc7a561ab0c0d2",
  Uint8Array [
    248,
    110,
    204,
    26,
    77,
    79,
    211,
    42,
    248,
    193,
    178,
    75,
    143,
    240,
    237,
    130,
    75,
    166,
    251,
    72,
    238,
    117,
    119,
    36,
    21,
    188,
    230,
    183,
    52,
    254,
    11,
    154,
    225,
    69,
    3,
    217,
    6,
    18,
    176,
    104,
    150,
    88,
    68,
    246,
    82,
    20,
    192,
    187,
    64,
    225,
    190,
    18,
    201,
    49,
    243,
    17,
    215,
    92,
    240,
    254,
    255,
    79,
    49,
    204,
    222,
    40,
    85,
    83,
  ],
]
`;
